chain/chain/src/chain.rs

```diff
@@ -1799,32 +1800,47 @@ impl Chain {
         Ok(())
     }
 
-    pub fn set_state_finalize(
+    pub fn start_set_state_finalize(
         &mut self,
         shard_id: ShardId,
         sync_hash: CryptoHash,
         num_parts: u64,
+        state_parts_task_scheduler: &dyn Fn(StatePartsMessage),
     ) -> Result<(), Error> {
         let shard_state_header = self.get_state_header(shard_id, sync_hash)?;
-        let mut height = shard_state_header.chunk_height_included();
         let state_root = shard_state_header.chunk_prev_state_root();
         let epoch_id = self.get_block_header(&sync_hash)?.epoch_id().clone();
 
+        let mut parts = Vec::new();
+
         for part_id in 0..num_parts {
             let key = StatePartKey(sync_hash, shard_id, part_id).try_to_vec()?;
             let part = self.store.owned_store().get(ColStateParts, &key)?.unwrap();
 
-            self.runtime_adapter.apply_state_part(
-                shard_id,
-                &state_root,
-                part_id,
-                num_parts,
-                &part,
-                &epoch_id,
-            )?;
+            parts.push(part);
```

Comment:
I think you cannot read `parts` here and send them over through `StatePartsMessage` because this will trigger OOM if the state is too large. The reason why the code needed to divide state to parts and apply it one by one is to limit memory usage.

To solve the issue, I think you will need to store a `Arc<Store>` in `StatePartsActor` and read the parts there.

Reply:
Right, I thought for some reason this have to do with downloading. Fixed